Add clarification for otel-js apps written in CJS and ESM

JamieDanielson opened the pull request:
Node.js has two module systems: [CommonJS(CJS) and ECMAScript modules(ESM)](https://nodejs.org/api/packages.html#determining-module-system). As ESM becomes more common in usage, we need to better document the current assumptions and requirements.

This is a start with the bare minimum, which is to specify the assumption that the app is running as CJS... as well as document the loader hook currently required for ESM apps. It seemed useful to start here based on issues like #4812.

More details will follow as we [finalize reference docs](https://github.com/open-telemetry/opentelemetry-js/pull/4876) in the otel-js repo.

trentm commented on content/en/docs/languages/js/getting-started/nodejs.md:
````diff
@@ -247,6 +251,17 @@ Listening for requests on http://localhost:8080
+$ node --experimental-loader=@opentelemetry/instrumentation/hook.mjs --require ./instrumentation.js app.js
+Listening for requests on http://localhost:8080
+```
+
````
Having the h3 section header for this unintentionally puts the subsequent content ("Open http://localhost:8080/rolldice in your web browser ...") under this section.

I don't have a great alternative suggestion. Perhaps the section could be replaced with a much shorter:

> (Note: If your application is written in JavaScript as ECMAScript Modules (ESM), or compiled to ESM from TypeScript, then a loader hook is required to properly support instrumentation. Use `node --experimental-loader=@opentelemetry/instrumentation/hook.mjs --require ./instrumentation.js app.js`. See [the coming reference link in opentelemetry-js/docs/] for details on ESM support.)

trentm commented on content/en/docs/languages/js/libraries.md:
```diff
@@ -19,6 +19,11 @@ For example,
 
+{{% alert title="Note" color="info" %}}
+Much of OpenTelemetry JS documentation is written assuming the compiled application is run as CommonJS.
+If the application runs as ESM, add the loader hook as specified in [Getting Started](/docs/languages/js/getting-started/nodejs/#instrumentation-for-ecmascript-modules).
```
Ah... in my suggestion above, I've removed this anchor. Perhaps the "see here for more info" could be the opentelemetry-js/doc/  document that is being added in https://github.com/open-telemetry/opentelemetry-js/pull/4876 instead?

theletterf commented on content/en/docs/languages/js/getting-started/nodejs.md:
```diff
@@ -24,6 +24,10 @@ Ensure that you have the following installed locally:
   TypeScript.
 
+{{% alert title="Note" color="info" %}}
+Much of OpenTelemetry JS documentation is written assuming the compiled application is run as CommonJS.
```
```suggestion
The OpenTelemetry documentation assumes that the compiled application is run as CommonJS.
```

Would it make sense to add a link to CommonJS? Are all Node developers familiar with it?

theletterf commented on content/en/docs/languages/js/getting-started/nodejs.md:
```diff
@@ -247,6 +251,17 @@ Listening for requests on http://localhost:8080
 
 {{% /tab %}} {{< /tabpane >}}
 
+### Instrumentation for ECMAScript Modules
```
```suggestion
### Instrumentation for ECMAScript modules
```

theletterf commented on content/en/docs/languages/js/getting-started/nodejs.md:
```diff
@@ -247,6 +251,17 @@ Listening for requests on http://localhost:8080
 
+### Instrumentation for ECMAScript Modules
+
+If your application is written in JavaScript as ESM, or compiled to ESM from TypeScript, then a loader hook is required to properly patch instrumentation.
+The custom hook for ESM instrumentation is `--experimental-loader=@opentelemetry/instrumentation/hook.mjs`.
+This flag must be passed to the `node` binary, which is often done as a startup command and/or in the `NODE_OPTIONS` environment variable.
```
```suggestion
If your application is written in JavaScript as ESM, or compiled to ESM
from TypeScript, a loader hook is required to patch instrumentation.
The custom hook for ESM instrumentation is
`--experimental-loader=@opentelemetry/instrumentation/hook.mjs`.
This flag must be passed to the `node` binary, which is often done
as a startup command or through the `NODE_OPTIONS` environment variable.
```

theletterf commented on content/en/docs/languages/js/libraries.md:
```diff
@@ -19,6 +19,11 @@ For example,
 the inbound HTTP requests.
 
+{{% alert title="Note" color="info" %}}
+Much of OpenTelemetry JS documentation is written assuming the compiled application is run as CommonJS.
```
```suggestion
The OpenTelemetry documentation assumes that the compiled application is run as CommonJS.
```

theletterf commented on content/en/docs/languages/js/instrumentation.md:
```diff
@@ -33,6 +33,10 @@ about manual instrumentation.
 library, follow the instructions here to adapt the process to your own code.
 
+{{% alert title="Note" color="info" %}}
+Much of OpenTelemetry JS documentation is written assuming the compiled application is run as CommonJS.
```
```suggestion
The OpenTelemetry documentation assumes that the compiled application is run as CommonJS.
```

theletterf commented on content/en/docs/languages/js/propagation.md:
```diff
@@ -7,6 +7,10 @@ cSpell:ignore: rolldice
 {{% docs/languages/propagation js %}}
 
+{{% alert title="Note" color="info" %}}
+Much of OpenTelemetry JS documentation is written assuming the compiled application is run as CommonJS.
```
```suggestion
The OpenTelemetry documentation assumes that the compiled application is run as CommonJS.
```

theletterf commented on content/en/docs/languages/js/resources.md:
```diff
@@ -15,6 +15,10 @@ Node.js SDK.
 files `package.json`, `app.js` and `tracing.js`.
 
+{{% alert title="Note" color="info" %}}
+Much of OpenTelemetry JS documentation is written assuming the compiled application is run as CommonJS.
```
```suggestion
The OpenTelemetry documentation assumes that the compiled application is run as CommonJS.
```

theletterf reviewed (changes requested):
Thanks! Some comments.

JamieDanielson commented:
I've moved this back to draft for now! It will be easier when we can link to another doc.

cartermp commented:
@JamieDanielson should this move back to review now that upstream is merged?

JamieDanielson commented:
> @JamieDanielson should this move back to review now that upstream is merged?

This completely fell off my radar, will double check for necessary changes and re-open for review!

svrnm commented:
is this PR still needed/something to work on?

JamieDanielson commented:
Sorry all let's close this for now 😅

JamieDanielson commented:
Hey I'm back! Opened a new PR to replace this #8666 as I wasn't sure what would be easiest to review. I incorporated the PR feedback and added the doc that now exists in the repo.
